lnpbp-0021.md

```diff
@@ -56,102 +56,111 @@ Interface specification is the following Contractum code:
 ```haskell
 -- # Defining main data structures
 
--- collectibles are usually scarse, so we limit their max number to 64k
 data ItemsCount :: U32
 
 -- each collectible item is unique and must have an id
-data TokenId :: U16
+data TokenId :: U32
 
 -- Collectibles are owned in fractions of a single item; here the owned
 -- fraction means `1/F`. Ownership of the full item is specified `F=1`;
 -- half of an item as `F=2` etc.
 data OwnedFraction :: U64
 
-data TxOut :: txid [Byte ^ 32], vout U16
-
 -- allocation of a single token or its fraction to some transaction output
 data Allocation :: TokenId, OwnedFraction
 
-data Engraving :: appliedTo TokenId, content Media
+data EngravingData :: 
+    appliedTo TokenId, 
+    content EmbeddedMedia
 
-data Media ::
-    type MimeType,
+data EmbeddedMedia ::
+    type RGBTypes.MediaType,
     data [Byte]
 
 data Attachment ::
-    type MimeType,
+    type RGBTypes.MediaType,
     digest: [U8 ^ 32] -- this can be any type of 32-byte hash, like SHA256(d), BLACKE3 etc
 
-data POR :: -- proof of reserves
-    reserves TxOut,
-    proof [Byte] -- schema-specific proof 
-
-data Token ::
+data TokenData ::
     id TokenId,
-    name [Ascii ^ 1..40],
-    details [Unicode ^ 40..256]?,
-    preview Media?, -- always embedded preview media < 64kb
-    media Attachment?, -- external media which is the main media for the token
+    ticker RGBTypes.Ticker?,
+    name RGBTypes.Name?,
+    details RGBTypes.Details?,
+    -- always-embedded preview media < 64kb
+    preview EmbeddedMedia?,
+    -- external media which is the main media for the token
+    media Attachment?,
     attachments { U8 -> ^ ..20 Attachment } -- auxiliary attachments by type (up to 20 attachments)
-    reserves POR? -- output containing locked bitcoins; how reserves are
-                  -- proved is a matter of a specific schema implementation
-
-data Denomination :: 
-    ticker [Ascii ^ 1..8],
-    name [Ascii ^ 1..40],
-    details [Unicode ^ 40..256]?,
-    contract [Unicode]??,
+    -- output containing locked bitcoins; how reserves are proved is a matter
+    -- of a specific schema implementation
+    reserves RGBTypes.PoR? 
 
  -- each attachment type is a mapping from attachment id 
  -- (used as `Token.attachments` keys) to a short Ascii string
  -- verbally explaining the type of the attachment for the UI
  -- (like "sample" etc).
-data AttachmentType :: id U8, description [Ascii ^ 1..20]
-
+data AttachmentType :: 
+    id U8, 
+    description [Ascii ^ 1..20]
 
 interface RGB21
-    global Name :: Denomination
-    global Tokens :: Token+
-    global Engravings :: Engraving+
+    global name :: RGBTypes.AssetNaming
+    
+    -- Contract text is separated from the name since it must not be
+    -- changeable by the issuer.
+    global terms :: RGBTypes.RicardianContract
+
+    -- Data for all issued tokens
+    global tokensData+ :: TokenData
+    global engravingData+ :: EngravingData
     global isFractional :: Bool
```

Comment:
I missed defining how much of an asset I can fractionate. For example, is it possible to have the nth part of an NFT/UDA?

Also, we can change the boolean to precision:

```suggestion
    global Fractionation :: RGBTypes.Precision
```

Reply:
Right, there should be a precision, which defines a base 10 U64 with a specific amount of zeroes, (e.g. 8 = 100,000,000), then the fraction should be a numerator =< than the denominator.

It should also be optional, which implies an absence of support for fractionalization.

Does this make sense?

Reply:
https://github.com/LNP-BP/LNPBPs/pull/154/commits/b54d9eac19a879df06fa88f1ac8ca8328eba998a#r1179658950